File: src/main/java/dev/vini2003/hammer/core/api/client/color/Color.java
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.core.api.client.color;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.nbt.NbtElement;
import net.minecraft.nbt.NbtLong;
import net.minecraft.network.PacketByteBuf;

/**
 * <p>A {@link Color} represents a color in the RGB color space.</p>
 *
 * <p>The following serialization methods are provided:</p>
 * <ul>
 *     <li>{@link #toJson(Color)} - from {@link Color} to {@link JsonElement}.</li>
 *     <li>{@link #toNbt(Color)} - from {@link Color} to {@link NbtElement}.</li>
 *     <li>{@link #toBuf(Color, PacketByteBuf)} - from {@link Color} to {@link PacketByteBuf}.</li>
 * </ul>

 * <ul>
 *     <li>{@link #fromJson(JsonElement)} from {@link JsonElement} to {@link Color}.</li>
 *     <li>{@link #fromNbt(NbtElement)} from {@link NbtElement} to {@link Color}.</li>
 *     <li>{@link #fromBuf(PacketByteBuf)} from {@link PacketByteBuf} to {@link Color}.</li>
 * </ul>
 */
public class Color {
	public static final Color DARK_RED = new Color(0.67F, 0.0F, 0.0F, 1.0F);
	public static final Color RED = new Color(1.0F, 0.33F, 0.33F, 1.0F);
	public static final Color GOLD = new Color(1.0F, 0.67F, 0.0F, 1.0F);
	public static final Color YELLOW = new Color(1.0F, 1.0F, 0.33F, 1.0F);
	public static final Color DARK_GREEN = new Color(0.0F, 0.67F, 0.0F, 1.0F);
	public static final Color GREEN = new Color(0.33F, 1.0F, 0.33F, 1.0F);
	public static final Color AQUA = new Color(0.33F, 1.0F, 1.0F, 1.0F);
	public static final Color DARK_AQUA = new Color(0.0F, 0.67F, 0.67F, 1.0F);
	public static final Color DARK_BLUE = new Color(0.0F, 0.0F, 0.67F, 1.0F);
	public static final Color BLUE = new Color(0.33F, 0.33F, 1.0F, 1.0F);
	public static final Color LIGHT_PURPLE = new Color(1.0F, 0.33F, 1.0F, 1.0F);
	public static final Color DARK_PURPLE = new Color(0.67F, 0.0F, 0.67F, 1.0F);
	public static final Color WHITE = new Color(1.0F, 1.0F, 1.0F, 1.0F);
	public static final Color GRAY = new Color(0.67F, 0.67F, 0.67F, 1.0F);
	public static final Color DARK_GRAY = new Color(0.33F, 0.33F, 0.33F, 1.0F);
	public static final Color BLACK = new Color(0.0F, 0.0F, 0.0F, 1.0F);
	
	private final float r;
	private final float g;
	private final float b;
	private final float a;
	
	/**
	 * Serializes a color to a {@link PacketByteBuf}.
	 * @param color The color to serialize.
	 * @param buf The buffer to serialize to.
	 * @return The buffer.
	 */
	public static PacketByteBuf toBuf(Color color, PacketByteBuf buf) {
		buf.writeLong(color.toRgba());
		
		return buf;
	}
	
	/**
	 * Deserializes a color from a {@link PacketByteBuf}.
	 * @param buf The buffer to deserialize from.
	 * @return The color.
	 */
	public static Color fromBuf(PacketByteBuf buf) {
		return new Color(buf.readLong());
	}
	
	/**
	 * Serializes a color to an {@link NbtCompound}.
	 * @param color The color.
	 * @return The serialized color.
	 */
	public static NbtElement toNbt(Color color) {
		return NbtLong.of(color.toRgba());
	}
	
	/**
	 * Deserializes a color from an {@link NbtCompound}.
	 * @param nbt The serialized color.
	 * @return The color.
	 */
	public static Color fromNbt(NbtElement nbt) {
		var rgba = ((NbtLong) nbt).longValue();
		
		return new Color(rgba);
	}
	
	/**
	 * Serializes a color to a {@link JsonElement}.
	 * @param color The color.
	 * @return The serialized color.
	 */
	public static JsonElement toJson(Color color) {
		var json = new JsonPrimitive(color.toRgba());
		
		return json;
	}
	
	/**
	 * Deserializes a color from a {@link JsonElement}.
	 * @param json The serialized color.
	 * @return The color.
	 */
	public static Color fromJson(JsonElement json) {
		var rgba = json.getAsLong();
		
		return new Color(rgba);
	}
	
	public Color(float r, float g, float b, float a) {
		this.r = r;
		this.g = g;
		this.b = b;
		this.a = a;
	}
	
	public Color(String rgbaHex) {
		this(Long.parseLong(rgbaHex, 16));
	}
	
	public Color(long rgba) {
		this.r = ((rgba >> 24) & 0xFF) / 255.0F;
		this.g = ((rgba >> 16) & 0xFF) / 255.0F;
		this.b = ((rgba >> 8) & 0xFF) / 255.0F;
		this.a = (rgba & 0xFF) / 255.0F;
	}
	
	public Color(int rgb) {
		this.r = ((rgb >> 16) & 0xFF) / 255.0F;
		this.g = ((rgb >> 8) & 0xFF) / 255.0F;
		this.b = (rgb & 0xFF) / 255.0F;
		this.a = 1.0F;
	}
	
	public float getR() {
		return r;
	}
	
	public float getG() {
		return g;
	}
	
	public float getB() {
		return b;
	}
	
	public float getA() {
		return a;
	}
	
	public long toRgba() {
		var r = (long) (this.r * 255.0F);
		var g = (long) (this.g * 255.0F);
		var b = (long) (this.b * 255.0F);
		var a = (long) (this.a * 255.0F);
		
		var rgba = r;
		rgba = (rgba << 8) + g;
		rgba = (rgba << 8) + b;
		rgba = (rgba << 8) + a;
		
		return rgba;
	}
	
	public int toRgb() {
		var r = (int) (this.r * 255.0F);
		var g = (int) (this.g * 255.0F);
		var b = (int) (this.b * 255.0F);
		
		var rgb = r;
		rgb = (rgb << 8) + g;
		rgb = (rgb << 8) + b;
		
		return rgb;
	}
}

File: hammer-core/src/main/kotlin/dev/vini2003/hammer/core/api/common/color/Color.kt
/*
 * MIT License
 *
 * Copyright (c) 2020 - 2022 vini2003
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package dev.vini2003.hammer.core.api.common.color

/**
 * A [Color] is an RGBA color, which contains red ([r]), green ([g]), blue ([b]) and alpha ([a]) components.
 */
class Color(
	/**
	 * Constructs a color.
	 *
	 * @param r the red percentage value.
	 * @param g the green percentage value.
	 * @param b the blue percentage value.
	 * @param a the alpha percentage value.
	 */
	val r: Float,
	val g: Float,
	val b: Float,
	val a: Float
) {
	/**
	 * Constructs a color.
	 *
	 * @param rgbaHex an RGBA hexadecimal string.
	 * @return the color.
	 */
	constructor(rgbaHex: String) : this(rgbaHex.toInt(16))
	
	/**
	 * Constructs a color.
	 *
	 * @param rgba an RGBA long.
	 * @return the color.
	 */
	constructor(rgba: Long) : this(((rgba shr 24) and 0xFF) / 255.0F, ((rgba shr 16) and 0xFF) / 255.0F, ((rgba shr 8) and 0xFF) / 255.0F, (rgba and 0xFF) / 255.0F)
	
	/**
	 * Constructs a color.
	 *
	 * @param rgb an RGB integer.
	 * @return the color.
	 */
	constructor(rgb: Int) : this(((rgb shr 16) and 0xFF) / 255.0F, ((rgb shr 8) and 0xFF) / 255.0F, ((rgb and 0xFF) / 255.0F), 1.0F)
	
	
	/**
	 * Returns this color as an RGB integer.
	 *
	 * @return the integer.
	 */
	fun toRGB(): Int {
		val r = (this.r * 255).toInt()
		val g = (this.g * 255).toInt()
		val b = (this.b * 255).toInt()
		
		var rgb = r
		
		rgb = (rgb shl 8) + g
		rgb = (rgb shl 8) + b
		
		return rgb
	}
	
	/**
	 * Returns this color as an RGBA long.
	 *
	 * @return the long.
	 */
	fun toRGBA(): Long {
		val r = (this.r * 255).toLong()
		val g = (this.g * 255).toLong()
		val b = (this.b * 255).toLong()
		val a = (this.b * 255).toLong()
		
		var rgba = r
		
		rgba = (rgba shl 8) + g
		rgba = (rgba shl 8) + b
		rgba = (rgba shl 8) + a
		
		return rgba
	}
	
	companion object {
		@JvmField
		val DARK_RED = Color(0.67F, 0.0F, 0.0F, 1.0F)
		
		@JvmField
		val RED = Color(1.0F, 0.33F, 0.33F, 1.0F)
		
		@JvmField
		val GOLD = Color(1.0F, 0.67F, 0.0F, 1.0F)
		
		@JvmField
		val YELLOW = Color(1.0F, 1.0F, 0.33F, 1.0F)
		
		@JvmField
		val DARK_GREEN = Color(0.0F, 0.67F, 0.0F, 1.0F)
		
		@JvmField
		val GREEN = Color(0.33F, 1.0F, 0.33F, 1.0F)
		
		@JvmField
		val AQUA = Color(0.33F, 1.0F, 1.0F, 1.0F)
		
		@JvmField
		val DARK_AQUA = Color(0.0F, 0.67F, 0.67F, 1.0F)
		
		@JvmField
		val DARK_BLUE = Color(0.0F, 0.0F, 0.67F, 1.0F)
		
		@JvmField
		val BLUE = Color(0.33F, 0.33F, 1.0F, 1.0F)
		
		@JvmField
		val LIGHT_PURPLE = Color(1.0F, 0.33F, 1.0F, 1.0F)
		
		@JvmField
		val DARK_PURPLE = Color(0.67F, 0.0F, 0.67F, 1.0F)
		
		@JvmField
		val WHITE = Color(1.0F, 1.0F, 1.0F, 1.0F)
		
		@JvmField
		val GRAY = Color(0.67F, 0.67F, 0.67F, 1.0F)
		
		@JvmField
		val DARK_GRAY = Color(0.33F, 0.33F, 0.33F, 1.0F)
		
		@JvmField
		val BLACK = Color(0.0F, 0.0F, 0.0F, 1.0F)
	}
}